src/core/main/Reflection/TypeGetMethodRecursivelyExtensions.cs

```diff
@@ -17,11 +17,50 @@ public static class TypeGetMethodRecursivelyExtensions
         /// <param name="argTypes">The type of the arguments you would use</param>
         /// <returns>A <see cref="System.Reflection.MethodInfo" /></returns>
         /// <exception cref="System.MissingMethodException">Thrown if the method could not be found</exception>
+        /// <exception cref="System.Reflection.AmbiguousMatchException">Thrown if two or more methods match the method definition</exception>
         public static MethodInfo GetMethodRecursively(this Type type, string methodName, params Type[] argTypes)
         {
             var typeInfo = type.GetTypeInfo();
-            var method = typeInfo.GetMethod(methodName, argTypes);
+            MethodInfo method = null;
+            
+            // Get all the methods that satisfy the method declaration
+            var methods = typeInfo
+                .GetMethods()
+                .Where(m => m.Name == methodName && m.GetParameters().Length == argTypes.Length)
+                .ToList()
+                .FindAll(m =>
+                {
+                    // Make sure that each parameter satisfies the method definition
+                    for (var parameterIndex = 0; parameterIndex < argTypes.Length; parameterIndex++)
+                    {
+                        if (argTypes[parameterIndex] == null)
+                        {
+                            var parameter = m.GetParameters()[parameterIndex].ParameterType;
+                            if (parameter.GetTypeInfo().IsValueType && Nullable.GetUnderlyingType(parameter) == null)
+                            {
+                                // If the argument given is null but the parameter is not nullable
+                                return false;
+                            }
+                        }
+                        else
+                        {
+                            if (m.GetParameters()[parameterIndex].ParameterType != argTypes[parameterIndex])
```

Comment:
I wonder if we actually need to check this slightly differently so that we allow people to provide `5` for a parameter that is actually of type `int?`.

E.g.

```csharp
if (m.GetParameters()[parameterIndex].ParameterType.GetTypeOrUnderlyingNullableType() != argTypes[parameterIndex])
```

Reply:
I agree that this needs to happen a bit differently. I created an issue for this problem: #110  However perhaps this should simply be fixed here.

Reply:
@pingvinen But is it actually desirable to do it in this method?
Think of this case:

```
public class SuperAwesomeBanana
{
    public void DoSomethingAwesome(int i) { /*Actually do something awesome*/ }
    public void DoSomethingAwesome(int? i) { /*Do something different but awesome*/ }
}
```

Although this probably is not a good code practice, this would create some problems if the underlying type is always considered, as these two methods would both satisfy a call to `someType.GetMethodRecursively("DoSomethingAwesome", typeof(int))`.

Reply:
Nice example :) I would argue that it is a code smell that should be fixed in the consumer's codebase, but it is still a case that we are now aware of and is easy to handle.

If we call our method with `5` then we should find both methods and therefore throw an `AmbiguousMatchException`. If we call our method with `null` then we will only find the nullable method.

Anywho... add a unit test that shows this behavior.

Reply:
@pingvinen Consider the `SuperAwesomeBanana` example again. What about a call like this:

```
typeof(SuperAwesomeBanana).GetMethodRecursively("DoSomethingAwesome", typeof(int?))
```

I find this a bit more tricky. I would say that the justification that `GetMethodRecursively` finds both overloaded methods when supplying the type: `typeof(int)`, is that it finds every method with the given name that always accepts parameters of the given types. However I would argue that the above mentioned call, where `typeof(int?)` is supplied should succeed by only finding one of the overloaded methods. This is because not all `int?`-values are applicable for both overloaded methods.

In other words my proposal is to change the code to the following:

```
var parameter = m.GetParameters()[parameterIndex].ParameterType;
if (parameter != argTypes[parameterIndex] &&
    parameter.GetTypeOrUnderlyingNullableType() != argTypes[parameterIndex])
{
    // If the argument and parameter does not match
    return false;
}
```

Reply:
> This is because not all int?-values are applicable for both overloaded methods.

I am not sure I follow...

I think that a direct call to `GetMethodRecursively` should return _the_ method matching the given signature. The compiler does not allow multiple methods with the same signature, so that should be fairly straightforward. And remember that `int != int?`.

The issue we are actually trying to solve is that if `InvokeMethodRecursively` is called with a `null` input, it has no way of telling the underlying call to `GetMethodRecursively` what the type of that argument is.

E.g.

```csharp
class DinosaursHadLegs
{
    public string DoMagic(string one, string two)
    {
    }

    public string DoMagic(string one, int two)
    {
    }

    public string DoMagic(string one, int? two)
    {
    }

    public string DoMagic(string one, NoTheyDidNot two)
    {
    }
}


var instance = new DinosaursHadLegs();

instance.InvokeMethodRecursively(nameof(DinosaursHadLegs.DoMagic), "one", "two");
// should have called DoMagic(string one, string two)

instance.InvokeMethodRecursively(nameof(DinosaursHadLegs.DoMagic), "one", 5);
// multiple matching signatures
// DoMagic(string one, int two)
// DoMagic(string one, int? two)


instance.InvokeMethodRecursively(nameof(DinosaursHadLegs.DoMagic), "one", null);
// multiple matching signatures
// DoMagic(string one, string two)
// DoMagic(string one, int? two)
// DoMagic(string one, NoTheyDidNot two)
```

I think `GetMethodRecursively` should treat `null` as `not sure about the type, but it must support being set to null`.